[ fix #4093 ] Bye bye Set

jespercockx opened the pull request:
This PR implements my proposed solution for #4093. It consists of the following changes:

- `Set` and `Prop` are removed as keywords, as are `Set0`, `Set1`, `Prop0`, ... and their unicode variants.
- `Agda.Primitive` now declares two new builtins `{-# BUILTIN SET Set #-}` and `{-# BUILTIN PROP Prop #-}`, which can be renamed when importing this module.
- All modules automatically start with an implicit statement `open import Agda.Primitive using (Set; Prop)`, unless the flag `--no-auto-import-primitive` is used.
- The scope checker has been extended to recognize suffixed versions of whatever names are bound to the builtins `SET` and `PROP`, so for example `Type0`, `Type1`, ... are now accepted when `Set` has been renamed to `Type`.
- The pretty-printer will also use the builtin names of `SET` and `PROP`, so the renamed versions are used in error messages and interactive commands.

If there is a consensus that this is a good design, I will add a changelog entry, some documentation, and a test case before merging.

jespercockx committed:
[ re #4093 ] Updated test cases

jespercockx committed:
[ #4093 ] Add new definition kind 'PrimitiveSort'

jespercockx committed:
[ #4093 ] Add optional suffix to Def in abstract syntax

jespercockx committed:
[ #4093 ] Add flag --no-auto-import-primitive (no effect yet)

jespercockx committed:
[ #4093 ] Lens for suffix of concrete name + some other utility functions

jespercockx committed:
[ #4093 ] Add new builtins SET and PROP

jespercockx committed:
[ #4093 ] Add 'sortKit' function for getting names of builtin sorts

jespercockx committed:
[ #4093 ] Add typing rules for new builtin sorts SET and PROP

jespercockx committed:
[ #4093 ] Add optional suffix to DefinedName in the scope checker

jespercockx committed:
[ refactor ] Use Lens' type synonym for lensAnameName and lensAmodName

jespercockx committed:
[ #4093 ] Parse in-scope name + suffix in scope checker

jespercockx committed:
[ #4093 ] Use builtins SET and PROP for printing sorts

jespercockx committed:
[ fix #4093 ] Make Set and Prop from keywords into builtins, enable auto-import of Agda.Primitive using (Set; Prop)

nad commented:
What happens if you rename `Set` to, say, `_X_X_`?

jespercockx commented:
This file is accepted:
```agda
{-# OPTIONS --no-auto-import-primitive #-}

open import Agda.Primitive renaming (Set to _X_X_)

test : _X_X₁_
test = _X_X_
```

guillaumebrunerie commented:
What if you rename `Set` to `Set` and `Prop` to `Set₁`? :smiley:

jespercockx commented:
```agda
{-# OPTIONS --no-auto-import-primitive --prop #-}

open import Agda.Primitive renaming (Set to Set; Prop to Set₁)

test : Set₁
test = Set
```
gives
```
Set₁ != Set₁
when checking that the expression Set has type Set₁
```
 :(

guillaumebrunerie commented:
Oh, sorry…
But, apart from that, I think that turning `Set` and `Prop` into pragmas that can be renamed is a great idea, thank you for making it possible!

MatthewDaggitt commented:
So next on the list is the ability to rename the `record` and `data` keywords right? Followed shortly by `where` and `module`... Then no Agda code will ever be able to be taken at face value 😄

jespercockx commented:
> So next on the list is the ability to rename the `record` and `data` keywords right? Followed shortly by `where` and `module`... Then no Agda code will ever be able to be taken at face value smile

Agda won't truly be configurable until we make it possible to rename `:` and `->` :P

guillaumebrunerie commented:
> > So next on the list is the ability to rename the `record` and `data` keywords right? Followed shortly by `where` and `module`... Then no Agda code will ever be able to be taken at face value smile
> 
> Agda won't truly be configurable until we make it possible to rename `:` and `->` :P

Yes! It should be possible to rename `:` to the Unicode colon, so that we can finally use a regular colon in syntax declarations!

nad commented:
What about the following code?
```agda
{-# OPTIONS --no-auto-import-primitive #-}

open import Agda.Primitive using (Prop)

postulate
  Set₁ : Prop

open import Agda.Primitive using (Set)

postulate
  A : Set₁
```

jespercockx commented:
> What about the following code?
> 
> ```agda
> {-# OPTIONS --no-auto-import-primitive #-}
> 
> open import Agda.Primitive using (Prop)
> 
> postulate
>   Set₁ : Prop
> 
> open import Agda.Primitive using (Set)
> 
> postulate
>   A : Set₁
> ```

The code is accepted if you add the `--prop` flag. The disambiguation favors the postulated `Set₁` over the suffixed version of the imported `Set`. Perhaps there should be a warning for this situation.

andreasabel commented:
> Perhaps there should be a warning for this situation.

Yes, we should have a bit of insurance against such obfuscation.

> If there is a consensus that this is a good design,

In the spirit of looking forward, I suggest to rename the `SET` builtin to `TYPE`.
```
{-# BUILTIN TYPE Set #-}
```

nad commented:
> In the spirit of looking forward, I suggest to rename the `SET` builtin to `TYPE`.
> 
> ```
> {-# BUILTIN TYPE Set #-}
> ```

I agree.

pnlph commented:
I was leaving [this comment](https://github.com/agda/agda/pull/4609#issuecomment-624175133) without seeing this discussion.

Could not be the default behaviour without implicit import and `{-# OPTIONS --auto-import-primitive #-}` to get the old behaviour? It is a bit strange having `Set` in scope in an empty module without options...

jespercockx commented:
> Could not be the default behaviour without implicit import and `{-# OPTIONS --auto-import-primitive #-}` to get the old behaviour? It is a bit strange having `Set` in scope in an empty module without options...

But currently `Set` is a keyword so it is always considered in scope. Not adding the implicit import would meant *all* old Agda files that mention `Set` would break, which is not something we'd want to do.

pnlph commented:
Sorry, I did not think about backward compatibility. Of course `Set` is practically everywhere... 🤦

In that case, I would upvote [this comment](https://github.com/agda/agda/issues/4093#issuecomment-621190147) from Andreas, `--bare` would be a short and descriptive new option.

Otherwise, would not be better `--no-auto-import-set` than `--no-auto-import-primitive`? Because the implicit import is only affecting `Set`, not the rest of the primitives like `Level`, `Setω`... Is that correct?

jespercockx commented:
> Sorry, I did not think about backward compatibility. Of course `Set` is practically everywhere... facepalm
> 
> In that case, I would upvote [this comment](https://github.com/agda/agda/issues/4093#issuecomment-621190147) from Andreas, `--bare` would be a short and descriptive new option.
> 
> Otherwise, would not be better `--no-auto-import-set` than `--no-auto-import-primitive`? Because the implicit import is only affecting `Set`, not the rest of the primitives like `Level`, `Setω`... Is that correct?

It affects `Set` and `Prop`.

laMudri commented:
Maybe, aiming for forward compatibility, the `--no-auto-import-primitive` flag should be more specific. We might in the future have code that uses `--no-auto-import-primitive` in order to rename `Set`, but then we use this mechanism to allow renaming of something else. Then code not using `--no-auto-import-primitive` inherits the shim, whereas code using it doesn't. Or, we might have to remember that `--no-auto-import-primitive` is just for `Prop` and `Set`, and introduce a more specific name for the flag for the other thing we end up renaming.

I would suggest having modules `Agda.Primitive.Set` and `Agda.Primitive.Prop`, and then options `--no-import-set` and `--no-import-prop`, or something like that.

jespercockx committed:
[ #4093 ] Changelog entry

jespercockx committed:
[ #4093 ] Update user manual

jespercockx committed:
[ #4093 ] Rename BUILTIN SET to BUILTIN TYPE

jespercockx committed:
[ #4093 ] Throw ambiguous name error when name could both be suffixed or non-suffixed

jespercockx committed:
[ #4093 ] Update LaTeX and HTML test output

jespercockx committed:
[ #4093 ] Add test cases

jespercockx committed:
[ #4093 ] Rename --no-auto-import-primitive to --no-import-sorts

jespercockx commented:
I've updated the PR with better handling of cases with ambiguous names, e.g. 
```agda
{-# OPTIONS --no-auto-import-primitive --prop #-}

open import Agda.Primitive renaming (Set to Set; Prop to Set₁)

test : Set₁
test = Set
```
now gives 
```
Ambiguous name Set₁. It could refer to any one of
  Set bound at AmbiguousSet1.agda:3,45-48
  Set₁ bound at
    AmbiguousSet1.agda:3,58-62
```
and
```agda
{-# OPTIONS --no-auto-import-primitive #-}

open import Agda.Primitive using (Prop)

postulate
  Set₁ : Prop

open import Agda.Primitive using (Set)

postulate
  A : Set₁
```
gives
```
Ambiguous name Set₁. It could refer to any one of
  Set bound at
    agda-default-include-path/Agda/Primitive.agda:13,18-21
  Set₁ bound at AmbiguousSet1b.agda:6,3-7
```
I also renamed the `SET` builtin to `TYPE`, and I've added entries to the changelog and user manual.

Finally, I renamed the `--no-auto-import-primitive` flag to `--no-import-sorts`, I hope this strikes a decent balance between brevity and clarity.

jespercockx commented:
This can be merged once the CI passes.

jespercockx committed:
[ refactor ] Clean up old code for Set and Prop

jespercockx commented:
My rebase messed things up, plz send help

jespercockx commented:
That looks better, let's hope CI passes this time.

andreasabel commented on src/full/Agda/Syntax/Scope/Monad.hs:
```diff
@@ -56,6 +57,7 @@ import Agda.Utils.Maybe
 import Agda.Utils.Monad
 import Agda.Utils.Null
 import Agda.Utils.Pretty
+import Agda.Utils.Suffix as C
```
This might have connections to #5706.
